crates/ra_hir/src/ty/pretty.rs

```diff
@@ -0,0 +1,48 @@
+use std::fmt;
+
+use crate::db::HirDatabase;
+
+pub struct PrettyPrintDisplay<'a, DB, T>(&'a DB, &'a T);
+
+impl<'a, DB, T> fmt::Display for PrettyPrintDisplay<'a, DB, T>
+where
+    DB: HirDatabase,
+    T: PrettyPrint,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.1.pp(self.0, f)
+    }
+}
+
+pub trait PrettyPrint {
+    fn pp(&self, db: &impl HirDatabase, f: &mut fmt::Formatter) -> fmt::Result;
+    fn pretty_print<'a, DB>(&'a self, db: &'a DB) -> PrettyPrintDisplay<'a, DB, Self>
```

Comment:
Let’s name this “display”? So that it works like Display for paths, but with an argument.

Reply:
Yeah, perhaps we can just name the trait “HirDisplay”?